File: Anime4K-Upscaler/Models/StreamOptimizeModels.swift
// Anime4K-Upscaler/Models/StreamOptimizeModels.swift
// Domain models for the Stream Optimize feature — configuration, enums, job, argument builder.

import Foundation
import SwiftUI
import Observation

// MARK: - Stream Encoder

/// Video encoder for stream optimization.
enum StreamEncoder: String, CaseIterable, Identifiable, Sendable {
    case hevcVideoToolbox = "hevc_videotoolbox"
    case h264VideoToolbox = "h264_videotoolbox"
    case svtAV1           = "libsvtav1"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .hevcVideoToolbox: return "HEVC (Hardware)"
        case .h264VideoToolbox: return "H.264 (Hardware)"
        case .svtAV1:           return "AV1 (Software)"
        }
    }

    var subtitle: String {
        switch self {
        case .hevcVideoToolbox: return "VideoToolbox GPU — Best balance of quality, size & compatibility"
        case .h264VideoToolbox: return "VideoToolbox GPU — Maximum device compatibility"
        case .svtAV1:           return "SVT-AV1 CPU — Best compression, modern players only"
        }
    }

    var symbolName: String {
        switch self {
        case .hevcVideoToolbox: return "bolt.fill"
        case .h264VideoToolbox: return "bolt"
        case .svtAV1:           return "cpu"
        }
    }

    /// Default quality value (HEVC/H264 use -q:v 0–100 higher=better, AV1 uses CRF 0–63 lower=better).
    var defaultQuality: Int {
        switch self {
        case .hevcVideoToolbox: return 65
        case .h264VideoToolbox: return 65
        case .svtAV1:           return 28
        }
    }

    var maxQuality: Int {
        switch self {
        case .hevcVideoToolbox: return 100
        case .h264VideoToolbox: return 100
        case .svtAV1:           return 63
        }
    }

    var qualityLabel: String {
        switch self {
        case .hevcVideoToolbox, .h264VideoToolbox:
            return "Quality (0–100, Higher = Better)"
        case .svtAV1:
            return "CRF (0–63, Lower = Better)"
        }
    }

    var usesCRF: Bool { self == .svtAV1 }

    /// Supported profiles for this encoder.
    var availableProfiles: [StreamProfile] {
        switch self {
        case .hevcVideoToolbox: return [.main, .main10]
        case .h264VideoToolbox: return [.high, .main, .baseline]
        case .svtAV1:           return [.main]  // AV1 main profile
        }
    }

    /// Default profile for this encoder.
    var defaultProfile: StreamProfile {
        switch self {
        case .hevcVideoToolbox: return .main10
        case .h264VideoToolbox: return .high
        case .svtAV1:           return .main
        }
    }

    /// Supported pixel formats for this encoder.
    var availablePixelFormats: [StreamPixelFormat] {
        switch self {
        case .hevcVideoToolbox: return [.p010le, .nv12]
        case .h264VideoToolbox: return [.nv12, .yuv420p]
        case .svtAV1:           return [.yuv420p10le, .yuv420p]
        }
    }

    /// Default pixel format for this encoder.
    var defaultPixelFormat: StreamPixelFormat {
        switch self {
        case .hevcVideoToolbox: return .p010le
        case .h264VideoToolbox: return .nv12
        case .svtAV1:           return .yuv420p10le
        }
    }
}

// MARK: - Stream Profile

/// Encoder profile options.
enum StreamProfile: String, CaseIterable, Identifiable, Sendable {
    case main10   = "main10"
    case main     = "main"
    case high     = "high"
    case baseline = "baseline"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .main10:   return "Main 10"
        case .main:     return "Main"
        case .high:     return "High"
        case .baseline: return "Baseline"
        }
    }

    var subtitle: String {
        switch self {
        case .main10:   return "10-bit color depth — Best HDR & gradient quality"
        case .main:     return "8-bit standard profile"
        case .high:     return "H.264 High — Best H.264 quality"
        case .baseline: return "Maximum compatibility (older devices)"
        }
    }
}

// MARK: - Stream Pixel Format

/// Pixel format options.
enum StreamPixelFormat: String, CaseIterable, Identifiable, Sendable {
    case p010le      = "p010le"
    case nv12        = "nv12"
    case yuv420p     = "yuv420p"
    case yuv420p10le = "yuv420p10le"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .p010le:      return "P010LE (10-bit)"
        case .nv12:        return "NV12 (8-bit HW)"
        case .yuv420p:     return "YUV420P (8-bit)"
        case .yuv420p10le: return "YUV420P10LE (10-bit)"
        }
    }
}

// MARK: - Stream Audio Mode

/// Audio handling strategy.
enum StreamAudioMode: String, CaseIterable, Identifiable, Sendable {
    case copy       = "copy"
    case aacTranscode = "aac"
    case aac128     = "aac_128"
    case aac192     = "aac_192"
    case aac256     = "aac_256"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .copy:       return "Copy (Passthrough)"
        case .aacTranscode: return "AAC (Auto Bitrate)"
        case .aac128:     return "AAC 128 kbps"
        case .aac192:     return "AAC 192 kbps"
        case .aac256:     return "AAC 256 kbps"
        }
    }

    var subtitle: String {
        switch self {
        case .copy:       return "Fastest — keeps original audio untouched"
        case .aacTranscode: return "Re-encode to AAC with FFmpeg default bitrate"
        case .aac128:     return "Good for spoken content & podcasts"
        case .aac192:     return "Balanced quality for music & movies"
        case .aac256:     return "High quality audio for critical listening"
        }
    }

    var symbolName: String {
        switch self {
        case .copy:          return "arrow.right.circle"
        case .aacTranscode:  return "waveform"
        case .aac128:        return "speaker.wave.1"
        case .aac192:        return "speaker.wave.2"
        case .aac256:        return "speaker.wave.3"
        }
    }

    /// Whether this mode copies audio without re-encoding.
    var isCopy: Bool { self == .copy }
}

// MARK: - Stream Subtitle Mode

/// Subtitle handling strategy for MP4 output.
enum StreamSubtitleMode: String, CaseIterable, Identifiable, Sendable {
    case movText = "mov_text"
    case copy    = "copy"
    case strip   = "strip"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .movText: return "MOV Text (MP4 Compatible)"
        case .copy:    return "Copy (Passthrough)"
        case .strip:   return "Strip All Subtitles"
        }
    }

    var subtitle: String {
        switch self {
        case .movText: return "Converts subtitles to MP4-native text — Best for streaming"
        case .copy:    return "Keeps original format — May fail in MP4 container"
        case .strip:   return "Removes all subtitle tracks"
        }
    }

    var symbolName: String {
        switch self {
        case .movText: return "text.bubble"
        case .copy:    return "arrow.right.circle"
        case .strip:   return "text.badge.minus"
        }
    }
}

// MARK: - Keyframe Interval

/// Keyframe interval presets for seeking performance.
enum KeyframeInterval: String, CaseIterable, Identifiable, Sendable {
    case oneSecond   = "1s"
    case twoSeconds  = "2s"
    case threeSeconds = "3s"
    case fiveSeconds = "5s"
    case tenSeconds  = "10s"

    var id: String { rawValue }

    /// Interval in seconds.
    var seconds: Int {
        switch self {
        case .oneSecond:    return 1
        case .twoSeconds:   return 2
        case .threeSeconds: return 3
        case .fiveSeconds:  return 5
        case .tenSeconds:   return 10
        }
    }

    var displayName: String {
        switch self {
        case .oneSecond:    return "1 second"
        case .twoSeconds:   return "2 seconds"
        case .threeSeconds: return "3 seconds"
        case .fiveSeconds:  return "5 seconds"
        case .tenSeconds:   return "10 seconds"
        }
    }

    var subtitle: String {
        switch self {
        case .oneSecond:    return "Instant seeking — Larger file size"
        case .twoSeconds:   return "Excellent seeking — Recommended for streaming"
        case .threeSeconds: return "Good seeking — Balanced"
        case .fiveSeconds:  return "Moderate seeking — Smaller file"
        case .tenSeconds:   return "Slow seeking — Smallest file"
        }
    }
}

// MARK: - Stream Optimize Configuration

/// User-configurable settings for stream optimization.
/// Defaults are tuned for Apple device streaming with quick seeking.
struct StreamOptimizeConfiguration: Sendable {
    var encoder: StreamEncoder         = .hevcVideoToolbox
    var quality: Int                   = 65
    var profile: StreamProfile         = .main10
    var pixelFormat: StreamPixelFormat = .p010le
    var audioMode: StreamAudioMode     = .copy
    var subtitleMode: StreamSubtitleMode = .movText
    var keyframeInterval: KeyframeInterval = .twoSeconds
    var faststart: Bool                = true
    var allowSWFallback: Bool          = true

    /// Production-ready defaults optimized for streaming delivery.
    static let `default` = StreamOptimizeConfiguration()
}

// MARK: - Stream Optimize Job

/// A single file being stream-optimized — tracks state, progress, and log output.
@MainActor @Observable
final class StreamOptimizeJob: Identifiable {
    let id: UUID
    let file: VideoFile
    let configuration: StreamOptimizeConfiguration

    var state: JobState = .idle
    var progress: Double = 0.0
    var currentFrame: Int = 0
    var currentTime: String = "00:00:00.00"
    var speed: String = "0.0x"
    var fps: String = "0"
    var outputURL: URL?
    var logLines: [String] = []
    var errorMessage: String?
    var startDate: Date?
    var endDate: Date?

    @ObservationIgnored var processHandle: Process?

    init(file: VideoFile, configuration: StreamOptimizeConfiguration, destinationDirectory: URL) {
        self.id = UUID()
        self.file = file
        self.configuration = configuration
        self.outputURL = destinationDirectory.appendingPathComponent("\(file.fileName)_streaming.mp4")
    }

    var elapsedTime: TimeInterval? {
        guard let start = startDate else { return nil }
        return (endDate ?? Date()).timeIntervalSince(start)
    }

    var formattedElapsedTime: String? {
        guard let elapsed = elapsedTime else { return nil }
        let hours = Int(elapsed) / 3600
        let minutes = (Int(elapsed) % 3600) / 60
        let seconds = Int(elapsed) % 60
        if hours > 0 {
            return String(format: "%02d:%02d:%02d", hours, minutes, seconds)
        } else {
            return String(format: "%02d:%02d", minutes, seconds)
        }
    }

    /// Append a log line (capped at 500 lines to prevent memory bloat).
    /// Uses in-place `removeSubrange` instead of allocating a new Array.
    func appendLog(_ line: String) {
        logLines.append(line)
        let overflow = logLines.count - 500
        if overflow > 0 {
            logLines.removeSubrange(0..<overflow)
        }
    }
}

// MARK: - Stream Optimize Argument Builder

/// Assembles FFmpeg arguments for stream optimization.
/// Produces output superb for streaming: frequent keyframes for quick seeking,
/// faststart moov atom, proper container metadata, and hardware-accelerated encoding.
struct StreamOptimizeArgumentBuilder {

    static func build(
        inputURL: URL,
        outputURL: URL,
        configuration: StreamOptimizeConfiguration
    ) -> [String] {
        var args: [String] = []
        args.reserveCapacity(32) // ~25-30 args typical; avoids backing-store reallocations

        // Global flags
        args.append(contentsOf: ["-nostdin", "-hide_banner", "-v", "error", "-stats"])
        args.append(contentsOf: ["-y"])
        args.append(contentsOf: ["-i", inputURL.path])

        // Stream mapping
        if configuration.subtitleMode == .strip {
            // Map video + audio only, strip subtitles
            args.append(contentsOf: ["-map", "0:v:0", "-map", "0:a?"])
        } else {
            // Map all streams
            args.append(contentsOf: ["-map", "0:v:0", "-map", "0:a?", "-map", "0:s?"])
        }

        // ─── VIDEO ENCODER ───
        switch configuration.encoder {
        case .hevcVideoToolbox:
            args.append(contentsOf: [
                "-c:v", "hevc_videotoolbox",
                "-q:v", "\(configuration.quality)",
                "-profile:v", configuration.profile.rawValue,
                "-pix_fmt", configuration.pixelFormat.rawValue
            ])

            if configuration.allowSWFallback {
                args.append(contentsOf: ["-allow_sw", "1"])
            }

        case .h264VideoToolbox:
            args.append(contentsOf: [
                "-c:v", "h264_videotoolbox",
                "-q:v", "\(configuration.quality)",
                "-profile:v", configuration.profile.rawValue,
                "-pix_fmt", configuration.pixelFormat.rawValue
            ])

            if configuration.allowSWFallback {
                args.append(contentsOf: ["-allow_sw", "1"])
            }

        case .svtAV1:
            args.append(contentsOf: [
                "-c:v", "libsvtav1",
                "-preset", "6",
                "-crf", "\(configuration.quality)",
                "-pix_fmt", configuration.pixelFormat.rawValue,
                "-svtav1-params", "tune=0"
            ])
        }

        // ─── KEYFRAME INTERVAL (Critical for quick seeking) ───
        // force_key_frames guarantees a keyframe exactly every N seconds,
        // independent of scene detection. This is essential for streaming
        // players that need frame-accurate seek points.
        let kfSeconds = configuration.keyframeInterval.seconds
        args.append(contentsOf: [
            "-force_key_frames", "expr:gte(t,n_forced*\(kfSeconds))"
        ])

        // ─── AUDIO ───
        switch configuration.audioMode {
        case .copy:
            args.append(contentsOf: ["-c:a", "copy"])
        case .aacTranscode:
            args.append(contentsOf: ["-c:a", "aac"])
        case .aac128:
            args.append(contentsOf: ["-c:a", "aac", "-b:a", "128k"])
        case .aac192:
            args.append(contentsOf: ["-c:a", "aac", "-b:a", "192k"])
        case .aac256:
            args.append(contentsOf: ["-c:a", "aac", "-b:a", "256k"])
        }

        // ─── SUBTITLES ───
        switch configuration.subtitleMode {
        case .movText:
            args.append(contentsOf: ["-c:s", "mov_text"])
        case .copy:
            args.append(contentsOf: ["-c:s", "copy"])
        case .strip:
            break // Already excluded via -map
        }

        // ─── STREAMING OPTIMIZATION ───
        if configuration.faststart {
            // Moves moov atom to the beginning of the file.
            // This is CRITICAL for streaming: players can start playback
            // immediately without downloading the entire file first.
            args.append(contentsOf: ["-movflags", "+faststart"])
        }

        // Output file
        args.append(outputURL.path)

        return args
    }
}

File: windows/src/models/StreamOptimizeModels.cpp
#include "StreamOptimizeModels.h"
#include "../utils/HardwareDetector.h"
#include <QFileInfo>
#include <QDir>

// MARK: - Stream Encoder Helper Functions

QString displayName(StreamEncoder encoder) {
    switch (encoder) {
        case StreamEncoder::HEVC_NVENC: return "HEVC (NVIDIA RTX)";
        case StreamEncoder::H264_NVENC: return "H.264 (NVIDIA RTX)";
        case StreamEncoder::HEVC_AMF:   return "HEVC (AMD)";
        case StreamEncoder::H264_AMF:   return "H.264 (AMD)";
        case StreamEncoder::HEVC_QSV:   return "HEVC (Intel)";
        case StreamEncoder::H264_QSV:   return "H.264 (Intel)";
        case StreamEncoder::SVT_AV1:     return "AV1 (Software)";
    }
    return QString();
}

QString subtitle(StreamEncoder encoder) {
    DeviceHardwareProfile profile = DeviceHardwareProfile::detect();
    switch (encoder) {
        case StreamEncoder::HEVC_NVENC: return QString("NVIDIA NVENC on %1 \u2014 Best balance of speed & quality").arg(profile.gpuName);
        case StreamEncoder::H264_NVENC: return QString("NVIDIA NVENC on %1 \u2014 Maximum device compatibility").arg(profile.gpuName);
        case StreamEncoder::HEVC_AMF:   return QString("AMD AMF on %1 \u2014 Best balance of speed & quality").arg(profile.gpuName);
        case StreamEncoder::H264_AMF:   return QString("AMD AMF on %1 \u2014 Maximum device compatibility").arg(profile.gpuName);
        case StreamEncoder::HEVC_QSV:   return QString("Intel QSV on %1 \u2014 Best balance of speed & quality").arg(profile.gpuName);
        case StreamEncoder::H264_QSV:   return QString("Intel QSV on %1 \u2014 Maximum device compatibility").arg(profile.gpuName);
        case StreamEncoder::SVT_AV1:     return QString("SVT-AV1 on %1-core CPU \u2014 Best compression for modern players").arg(profile.cpuCoreCount);
    }
    return QString();
}

QString symbolName(StreamEncoder encoder) {
    switch (encoder) {
        case StreamEncoder::HEVC_NVENC:
        case StreamEncoder::HEVC_AMF:
        case StreamEncoder::HEVC_QSV:
            return "bolt.fill";
        case StreamEncoder::H264_NVENC:
        case StreamEncoder::H264_AMF:
        case StreamEncoder::H264_QSV:
            return "bolt";
        case StreamEncoder::SVT_AV1:
            return "cpu";
    }
    return QString();
}

int defaultQuality(StreamEncoder encoder) {
    switch (encoder) {
        case StreamEncoder::HEVC_NVENC:
        case StreamEncoder::H264_NVENC:
        case StreamEncoder::HEVC_AMF:
        case StreamEncoder::H264_AMF:
        case StreamEncoder::HEVC_QSV:
        case StreamEncoder::H264_QSV:
            return 65;
        case StreamEncoder::SVT_AV1:
            return 28;
    }
    return 65;
}

int maxQuality(StreamEncoder encoder) {
    switch (encoder) {
        case StreamEncoder::HEVC_NVENC:
        case StreamEncoder::H264_NVENC:
        case StreamEncoder::HEVC_AMF:
        case StreamEncoder::H264_AMF:
        case StreamEncoder::HEVC_QSV:
        case StreamEncoder::H264_QSV:
            return 100;
        case StreamEncoder::SVT_AV1:
            return 63;
    }
    return 100;
}

QString qualityLabel(StreamEncoder encoder) {
    if (encoder == StreamEncoder::SVT_AV1) {
        return "CRF (0\u201363, Lower = Better)";
    }
    return "Quality (0\u2013100, Higher = Better)";
}

bool usesCRF(StreamEncoder encoder) {
    return encoder == StreamEncoder::SVT_AV1;
}

// MARK: - Stream Profile Helper Functions

QString displayName(StreamProfile profile) {
    switch (profile) {
        case StreamProfile::Main10:   return "Main 10";
        case StreamProfile::Main:     return "Main";
        case StreamProfile::High:     return "High";
        case StreamProfile::Baseline: return "Baseline";
    }
    return QString();
}

QString subtitle(StreamProfile profile) {
    switch (profile) {
        case StreamProfile::Main10:   return "10-bit color depth \u2014 Best HDR & gradient quality";
        case StreamProfile::Main:     return "8-bit standard profile";
        case StreamProfile::High:     return "H.264 High \u2014 Best H.264 quality";
        case StreamProfile::Baseline: return "Maximum compatibility (older devices)";
    }
    return QString();
}

QString profileValue(StreamProfile profile) {
    switch (profile) {
        case StreamProfile::Main10:   return "main10";
        case StreamProfile::Main:     return "main";
        case StreamProfile::High:     return "high";
        case StreamProfile::Baseline: return "baseline";
    }
    return QString();
}

QVector<StreamProfile> availableProfiles(StreamEncoder encoder) {
    switch (encoder) {
        case StreamEncoder::HEVC_NVENC:
        case StreamEncoder::HEVC_AMF:
        case StreamEncoder::HEVC_QSV:
            return { StreamProfile::Main, StreamProfile::Main10 };
        case StreamEncoder::H264_NVENC:
        case StreamEncoder::H264_AMF:
        case StreamEncoder::H264_QSV:
            return { StreamProfile::High, StreamProfile::Main, StreamProfile::Baseline };
        case StreamEncoder::SVT_AV1:
            return { StreamProfile::Main };
    }
    return {};
}

StreamProfile defaultProfile(StreamEncoder encoder) {
    switch (encoder) {
        case StreamEncoder::HEVC_NVENC:
        case StreamEncoder::HEVC_AMF:
        case StreamEncoder::HEVC_QSV:
            return StreamProfile::Main10;
        case StreamEncoder::H264_NVENC:
        case StreamEncoder::H264_AMF:
        case StreamEncoder::H264_QSV:
            return StreamProfile::High;
        case StreamEncoder::SVT_AV1:
            return StreamProfile::Main;
    }
    return StreamProfile::Main;
}

// MARK: - Stream Pixel Format Helper Functions

QString displayName(StreamPixelFormat pixFmt) {
    switch (pixFmt) {
        case StreamPixelFormat::P010LE:      return "P010LE (10-bit)";
        case StreamPixelFormat::NV12:        return "NV12 (8-bit HW)";
        case StreamPixelFormat::YUV420P:     return "YUV420P (8-bit)";
        case StreamPixelFormat::YUV420P10LE: return "YUV420P10LE (10-bit)";
    }
    return QString();
}

QString pixelFormatValue(StreamPixelFormat pixFmt) {
    switch (pixFmt) {
        case StreamPixelFormat::P010LE:      return "p010le";
        case StreamPixelFormat::NV12:        return "nv12";
        case StreamPixelFormat::YUV420P:     return "yuv420p";
        case StreamPixelFormat::YUV420P10LE: return "yuv420p10le";
    }
    return QString();
}

QVector<StreamPixelFormat> availablePixelFormats(StreamEncoder encoder) {
    switch (encoder) {
        case StreamEncoder::HEVC_NVENC:
        case StreamEncoder::HEVC_AMF:
        case StreamEncoder::HEVC_QSV:
            return { StreamPixelFormat::P010LE, StreamPixelFormat::NV12 };
        case StreamEncoder::H264_NVENC:
        case StreamEncoder::H264_AMF:
        case StreamEncoder::H264_QSV:
            return { StreamPixelFormat::NV12, StreamPixelFormat::YUV420P };
        case StreamEncoder::SVT_AV1:
            return { StreamPixelFormat::YUV420P10LE, StreamPixelFormat::YUV420P };
    }
    return {};
}

StreamPixelFormat defaultPixelFormat(StreamEncoder encoder) {
    switch (encoder) {
        case StreamEncoder::HEVC_NVENC:
        case StreamEncoder::HEVC_AMF:
        case StreamEncoder::HEVC_QSV:
            return StreamPixelFormat::P010LE;
        case StreamEncoder::H264_NVENC:
        case StreamEncoder::H264_AMF:
        case StreamEncoder::H264_QSV:
            return StreamPixelFormat::NV12;
        case StreamEncoder::SVT_AV1:
            return StreamPixelFormat::YUV420P10LE;
    }
    return StreamPixelFormat::NV12;
}

// MARK: - Stream Audio Mode Helper Functions

QString displayName(StreamAudioMode mode) {
    switch (mode) {
        case StreamAudioMode::Copy:         return "Copy (Passthrough)";
        case StreamAudioMode::AACTranscode: return "AAC (Auto Bitrate)";
        case StreamAudioMode::AAC128:       return "AAC 128 kbps";
        case StreamAudioMode::AAC192:       return "AAC 192 kbps";
        case StreamAudioMode::AAC256:       return "AAC 256 kbps";
    }
    return QString();
}

QString subtitle(StreamAudioMode mode) {
    switch (mode) {
        case StreamAudioMode::Copy:         return "Fastest \u2014 keeps original audio untouched";
        case StreamAudioMode::AACTranscode: return "Re-encode to AAC with FFmpeg default bitrate";
        case StreamAudioMode::AAC128:       return "Good for spoken content & podcasts";
        case StreamAudioMode::AAC192:       return "Balanced quality for music & movies";
        case StreamAudioMode::AAC256:       return "High quality audio for critical listening";
    }
    return QString();
}

QString symbolName(StreamAudioMode mode) {
    switch (mode) {
        case StreamAudioMode::Copy:         return "arrow.right.circle";
        case StreamAudioMode::AACTranscode: return "waveform";
        case StreamAudioMode::AAC128:       return "speaker.wave.1";
        case StreamAudioMode::AAC192:       return "speaker.wave.2";
        case StreamAudioMode::AAC256:       return "speaker.wave.3";
    }
    return QString();
}

bool isCopy(StreamAudioMode mode) {
    return mode == StreamAudioMode::Copy;
}

// MARK: - Stream Subtitle Mode Helper Functions

QString displayName(StreamSubtitleMode mode) {
    switch (mode) {
        case StreamSubtitleMode::MovText: return "MOV Text (MP4 Compatible)";
        case StreamSubtitleMode::Copy:    return "Copy (Passthrough)";
        case StreamSubtitleMode::Strip:   return "Strip All Subtitles";
    }
    return QString();
}

QString subtitle(StreamSubtitleMode mode) {
    switch (mode) {
        case StreamSubtitleMode::MovText: return "Converts subtitles to MP4-native text \u2014 Best for streaming";
        case StreamSubtitleMode::Copy:    return "Keeps original format \u2014 May fail in MP4 container";
        case StreamSubtitleMode::Strip:   return "Removes all subtitle tracks";
    }
    return QString();
}

QString symbolName(StreamSubtitleMode mode) {
    switch (mode) {
        case StreamSubtitleMode::MovText: return "text.bubble";
        case StreamSubtitleMode::Copy:    return "arrow.right.circle";
        case StreamSubtitleMode::Strip:   return "text.badge.minus";
    }
    return QString();
}

// MARK: - Keyframe Interval Helper Functions

int seconds(KeyframeInterval interval) {
    switch (interval) {
        case KeyframeInterval::OneSecond:    return 1;
        case KeyframeInterval::TwoSeconds:   return 2;
        case KeyframeInterval::ThreeSeconds: return 3;
        case KeyframeInterval::FiveSeconds:  return 5;
        case KeyframeInterval::TenSeconds:   return 10;
    }
    return 2;
}

QString displayName(KeyframeInterval interval) {
    switch (interval) {
        case KeyframeInterval::OneSecond:    return "1 second";
        case KeyframeInterval::TwoSeconds:   return "2 seconds";
        case KeyframeInterval::ThreeSeconds: return "3 seconds";
        case KeyframeInterval::FiveSeconds:  return "5 seconds";
        case KeyframeInterval::TenSeconds:   return "10 seconds";
    }
    return QString();
}

QString subtitle(KeyframeInterval interval) {
    switch (interval) {
        case KeyframeInterval::OneSecond:    return "Instant seeking \u2014 Larger file size";
        case KeyframeInterval::TwoSeconds:   return "Excellent seeking \u2014 Recommended for streaming";
        case KeyframeInterval::ThreeSeconds: return "Good seeking \u2014 Balanced";
        case KeyframeInterval::FiveSeconds:  return "Moderate seeking \u2014 Smaller file";
        case KeyframeInterval::TenSeconds:   return "Slow seeking \u2014 Smallest file";
    }
    return QString();
}

// MARK: - Stream Optimize Configuration

StreamOptimizeConfiguration StreamOptimizeConfiguration::defaultConfig() {
    return StreamOptimizeConfiguration();
}

// MARK: - Stream Optimize Job

StreamOptimizeJob::StreamOptimizeJob(const VideoFile& f, const StreamOptimizeConfiguration& c, const QString& destinationDirectory, QObject* parent)
    : QObject(parent), file(f), configuration(c) {
    id = QUuid::createUuid();
    outputPath = QDir(destinationDirectory).filePath(QString("%1_streaming.mp4").arg(file.fileName));
}

double StreamOptimizeJob::elapsedTime() const {
    if (!startDate.isValid()) return 0.0;
    QDateTime end = endDate.isValid() ? endDate : QDateTime::currentDateTime();
    return startDate.msecsTo(end) / 1000.0;
}

QString StreamOptimizeJob::formattedElapsedTime() const {
    double elapsed = elapsedTime();
    int hours = static_cast<int>(elapsed) / 3600;
    int minutes = (static_cast<int>(elapsed) % 3600) / 60;
    int seconds = static_cast<int>(elapsed) % 60;
    if (hours > 0) {
        return QString("%1:%2:%3")
            .arg(hours, 2, 10, QChar('0'))
            .arg(minutes, 2, 10, QChar('0'))
            .arg(seconds, 2, 10, QChar('0'));
    } else {
        return QString("%1:%2")
            .arg(minutes, 2, 10, QChar('0'))
            .arg(seconds, 2, 10, QChar('0'));
    }
}

void StreamOptimizeJob::appendLog(const QString& line) {
    logLines.append(line);
    int overflow = logLines.count() - 500;
    if (overflow > 0) {
        logLines.erase(logLines.begin(), logLines.begin() + overflow);
    }
}

// MARK: - Stream Optimize Argument Builder

QStringList StreamOptimizeArgumentBuilder::build(
    const QString& inputPath,
    const QString& outputPath,
    const StreamOptimizeConfiguration& configuration
) {
    QStringList args;
    args.append({"-nostdin", "-hide_banner", "-v", "error", "-stats"});
    args.append("-y");
    args.append({"-i", inputPath});

    // Subtitle mapping
    if (configuration.subtitleMode == StreamSubtitleMode::Strip) {
        args.append({"-map", "0:v:0", "-map", "0:a?"});
    } else {
        args.append({"-map", "0:v:0", "-map", "0:a?", "-map", "0:s?"});
    }

    // Video Encoder
    switch (configuration.encoder) {
        case StreamEncoder::HEVC_NVENC:
            args.append({"-c:v", "hevc_nvenc",
                         "-preset", "p4",
                         "-tune", "hq",
                         "-rc", "vbr",
                         "-cq", QString::number(configuration.quality),
                         "-b:v", "0",
                         "-profile:v", profileValue(configuration.profile),
                         "-pix_fmt", pixelFormatValue(configuration.pixelFormat)});
            break;

        case StreamEncoder::H264_NVENC:
            args.append({"-c:v", "h264_nvenc",
                         "-preset", "p4",
                         "-tune", "hq",
                         "-rc", "vbr",
                         "-cq", QString::number(configuration.quality),
                         "-b:v", "0",
                         "-profile:v", profileValue(configuration.profile),
                         "-pix_fmt", pixelFormatValue(configuration.pixelFormat)});
            break;

        case StreamEncoder::HEVC_AMF:
            args.append({"-c:v", "hevc_amf",
                         "-quality", "quality",
                         "-rc", "cqp",
                         "-qp_i", QString::number(configuration.quality),
                         "-qp_p", QString::number(configuration.quality),
                         "-profile:v", profileValue(configuration.profile),
                         "-pix_fmt", pixelFormatValue(configuration.pixelFormat)});
            break;

        case StreamEncoder::H264_AMF:
            args.append({"-c:v", "h264_amf",
                         "-quality", "quality",
                         "-rc", "cqp",
                         "-qp_i", QString::number(configuration.quality),
                         "-qp_p", QString::number(configuration.quality),
                         "-profile:v", profileValue(configuration.profile),
                         "-pix_fmt", pixelFormatValue(configuration.pixelFormat)});
            break;

        case StreamEncoder::HEVC_QSV:
            args.append({"-c:v", "hevc_qsv",
                         "-preset", "medium",
                         "-global_quality", QString::number(configuration.quality),
                         "-profile:v", profileValue(configuration.profile),
                         "-pix_fmt", pixelFormatValue(configuration.pixelFormat)});
            break;

        case StreamEncoder::H264_QSV:
            args.append({"-c:v", "h264_qsv",
                         "-preset", "medium",
                         "-global_quality", QString::number(configuration.quality),
                         "-profile:v", profileValue(configuration.profile),
                         "-pix_fmt", pixelFormatValue(configuration.pixelFormat)});
            break;

        case StreamEncoder::SVT_AV1:
            args.append({"-c:v", "libsvtav1",
                         "-preset", "6",
                         "-crf", QString::number(configuration.quality),
                         "-pix_fmt", pixelFormatValue(configuration.pixelFormat),
                         "-svtav1-params", "tune=0"});
            break;
    }

    // Keyframe interval forced
    int kfSec = seconds(configuration.keyframeInterval);
    args.append({"-force_key_frames", QString("expr:gte(t,n_forced*%1)").arg(kfSec)});

    // Audio encoding
    switch (configuration.audioMode) {
        case StreamAudioMode::Copy:
            args.append({"-c:a", "copy"});
            break;
        case StreamAudioMode::AACTranscode:
            args.append({"-c:a", "aac"});
            break;
        case StreamAudioMode::AAC128:
            args.append({"-c:a", "aac", "-b:a", "128k"});
            break;
        case StreamAudioMode::AAC192:
            args.append({"-c:a", "aac", "-b:a", "192k"});
            break;
        case StreamAudioMode::AAC256:
            args.append({"-c:a", "aac", "-b:a", "256k"});
            break;
    }

    // Subtitle encoding
    switch (configuration.subtitleMode) {
        case StreamSubtitleMode::MovText:
            args.append({"-c:s", "mov_text"});
            break;
        case StreamSubtitleMode::Copy:
            args.append({"-c:s", "copy"});
            break;
        case StreamSubtitleMode::Strip:
            break;
    }

    // Streaming optimization
    if (configuration.faststart) {
        args.append({"-movflags", "+faststart"});
    }

    args.append(outputPath);
    return args;
}
